### PR TITLE
Add the sentence position prediction objective

### PR DESCRIPTION
**Pull request by armancohan**
- Data generation shuffles sentences 
- Model adds a new objective to predict the correct sentence position

### CODE REVIEW

File: pytorch_transformers/modeling_bert.py [hunk 1]
```diff
@@ -779,21 +799,55 @@ def tie_weights(self):
...
 
-        if masked_lm_labels is not None and next_sentence_label is not None:
+        # Add 1 dimensiton to use for TimeDistributed feed forward, shape = [1, N, D]
+        sep_output_scores = sep_output_scores.unsqueeze(0)
+
+        segment_prediction_scores = self.segment_classifier(sep_output_scores)
+
+        if original_segment_ids is not None:
+            # select labels that correspond to the actual SEP tokens, shape = [N] where N=number of SEP tokens in the batch
+            segment_labels = original_segment_ids[sep_sentences_mask]
```
Comment (ibeltagy) [inline pytorch_transformers/modeling_bert.py]: indexing a tensor by another tensor doesn't work on TPU. Will address this in a separate PR

File: pytorch_transformers/modeling_bert.py [hunk 2]
```diff
@@ -523,6 +527,20 @@ def forward(self, pooled_output):
         return seq_relationship_score
 
 
+class BertSentClassifierPreTrainingHeads(nn.Module):
+    """ Sentence classifier for the shuffling objective """
+    def __init__(self, config):
+        super(BertSentClassifierPreTrainingHeads, self).__init__()
+        self.sent_classifier = TimeDistributed(nn.Linear(config.hidden_size, config.max_num_sents))
```
Comment (ibeltagy) [inline pytorch_transformers/modeling_bert.py]: is `TimeDistributed` needed here? I thought `nn.Linear` can work on tensors of arbitrary number of dimensions

File: examples/lm_finetuning/pregenerate_training_data.py [hunk 3]
```diff
@@ -82,6 +95,32 @@ def __exit__(self, exc_type, exc_val, traceback):
             self.temp_dir.cleanup()
 
 
+def truncate_seq(sents: List[str], max_num_tokens: int):
+    """ Truncates a sequence of sentences to a maximum sequence length 
+    Chooses the list with the maximum number of tokens and truncates from that until the total length is within
```
[the remark this answers is not in the document]
Comment (armancohan) [reply examples/lm_finetuning/pregenerate_training_data.py]: sentence = list of tokens
will clarify

File: pytorch_transformers/modeling_bert.py (hunk 2 above)
Comment (armancohan) [reply pytorch_transformers/modeling_bert.py]: Yes, good point. Doesn't seem necessary.

File: examples/lm_finetuning/pregenerate_training_data.py [hunk 4]
```diff
@@ -195,72 +236,86 @@ def create_instances_from_document(
...
+                chunks_shuffled = [chunks[shuffled_idx] for shuffled_idx in chunk_ids_shuffled]
+
+                # tokens: "['[CLS]', w1, w2, '[SEP]', w3, w4, '[SEP]', w5, w6, w7, '[SEP]'"
+                tokens = ["[CLS]"] + [e for segment_tokens in chunks_shuffled for e in segment_tokens + ["[SEP]"]]
+
+                # segment_ids: [0, 0, 0, 1, 1, 1, 2, 2, 2, 2, 3, 3]
+                # orig_segment_ids : [1, 1, 1, 3, 3, 0, 0, 0, 2, 2, 2, 2 ]
+                orig_segment_ids = []
+                segment_ids = []
+                for j, tokens_a in enumerate(chunks):
```
Comment (ibeltagy) [inline examples/lm_finetuning/pregenerate_training_data.py]: does this need to be `for j, tokens_a in enumerate(chunks_shuffled)` ?
is it `orig_segment_ids` or `segment_ids` that needs to be aligned with `tokens` ?
for example, 
```
chunk_ids_shuffled = [1, 0]
chunks = [[w1, w2, w3], [z1, z2, z3, z4]] 
chunks_shuffled = [[z1, z2, z3, z4], [w1, w2, w3]]  
tokens =CLS, z1, z2, z3, z4, SEP w1, w2, w3, SEP
orig_segment_ids = 1, 1, 1, 1, 1, 0, 0, 0, 0, 0   # current values, not aligned with tokens
orig_segment_ids = 1, 1, 1, 1, 1, 1, 0, 0, 0, 0   # values if looped over chunks_shuffled -> aligned with tokens
```
